packages/node/src/channels.ts

```diff
@@ -0,0 +1,328 @@
+import { legacy } from "@counterfactual/cf.js";
+import {
+  Address,
+  AppInstanceInfo,
+  AssetType,
+  Node
+} from "@counterfactual/common-types";
+import { ethers } from "ethers";
+
+import { IStoreService } from "./service-interfaces";
+
+import Nonce = legacy.utils.Nonce;
+import FreeBalance = legacy.utils.FreeBalance;
+
+/**
+ * The schema of a channel is below.
+ * The following Channels class encapsulates the state and persistence of all
+ * the channels in a Node.
+ */
+
+/**
+ * The fully expanded schema for a channel:
+ * multisigAddress: {
+ *  multisigAddress: Address,
+ *  multisigOwners: Address[],
+ *  appsNonce: Nonce,
+ *  appInstances: Map<AppInstanceID,
+ *    appInstance: {
+ *      id: string,
+ *      appId: Address,
+ *      abiEncodings: {
+ *        stateEncoding: string,
+ *        actionEncoding: string
+ *      },
+ *      appState: any,
+ *      localNonce: Nonce,
+ *      dependencyNonce: Nonce,
+ *      timeout: BigNumber,
+ *      asset: {
+ *        assetType: AssetType,
+ *        token?: Address
+ *      },
+ *      deposits: Map<Address, BigNumber>
+ *    }
+ *  },
+ *  proposedAppInstances: same schema as appInstances,
+ *  freeBalances: Map<AssetType,
+ *    freeBalance: {
+ *      alice: Address,
+ *      aliceBalance: BigNumber,
+ *      bob: Address,
+ *      bobBalance: BigNumber,
+ *      uniqueId: number,
+ *      localNonce: number,
+ *      timeout: number,
+ *      dependencyNonce: {
+ *        isSet: boolean,
+ *        salt: string,
+ *        nonceValue: number
+ *      }
+ *    }
+ *  }
+ * }
+ */
+
+/**
+ * This class is only a type implementation of a channel schema for the
+ * purposes of updating and retrieving a channel's state from the store.
+ *
+ * An instance is by itself stateless and effectively reflects the state of an
+ * according channel in the store.
+ */
+class Channel {
+  constructor(
+    readonly multisigAddress: Address,
+    readonly multisigOwners: Address[],
+    readonly appsNonce: Nonce = new Nonce(true, 0, 0),
+    readonly freeBalances: {
+      [assetType: number]: FreeBalance;
+    } = Channel.initialFreeBalances(multisigOwners, appsNonce),
+    readonly appInstances: {
+      [appInstanceId: string]: AppInstanceInfo;
+    } = {},
+    readonly proposedAppInstances: {
+      [appInstanceId: string]: AppInstanceInfo;
+    } = {}
+  ) {}
+
+  static initialFreeBalances(
+    multisigOwners: Address[],
+    initialAppsNonce: Nonce
+  ): {
+    [assetType: number]: FreeBalance;
+  } {
+    // TODO: extend to all asset types
+    const ethFreeBalance = new FreeBalance(
+      multisigOwners[0],
+      ethers.utils.bigNumberify("0"),
```

Comment:
@cf19drofxots Might be interesting to have a constant for this. I've noticed we use it several times.